symengine/logic.h

```diff
@@ -96,6 +96,128 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
     return make_rcp<Piecewise>(std::move(vec));
 }
 
+class And : public Boolean
+{
+public:
+    set_boolean args;
+
+public:
+    IMPLEMENT_TYPEID(AND)
+    And(const set_boolean &s);
+    bool is_canonical(const set_boolean &container);
+    //! \return the hash
+    std::size_t __hash__() const;
+    virtual vec_basic get_args() const;
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+};
+
+class Or : public Boolean
+{
+public:
+    set_boolean args;
+
+public:
+    IMPLEMENT_TYPEID(OR)
+    Or(const set_boolean &s);
+    bool is_canonical(const set_boolean &container);
+    //! \return the hash
+    std::size_t __hash__() const;
+    virtual vec_basic get_args() const;
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+};
+
+class Not : public Boolean
+{
+public:
+    RCP<const Boolean> arg;
+
+public:
+    IMPLEMENT_TYPEID(NOT)
+    Not(const RCP<const Boolean> &s);
+    //! \return the hash
+    std::size_t __hash__() const;
+    virtual vec_basic get_args() const;
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+};
+
+RCP<const Boolean> logical_not(const RCP<const Boolean> &s);
+
+template <typename caller>
+RCP<const Boolean> and_or(const set_boolean &s, const bool &op_x_notx)
+{
+    set_boolean args;
+    for (auto &a : s) {
+        if (is_a<BooleanAtom>(*a)) {
+            auto val = static_cast<const BooleanAtom &>(*a).get_val();
+            if (val == op_x_notx)
+                return boolean(op_x_notx);
+            else
+                continue;
+        }
+        args.insert(a);
+    }
+    set_boolean flattened_args;
+    for (auto &a : args) {
+        if (is_a<caller>(*a)) {
+            const caller &to_insert = static_cast<const caller &>(*a);
+            flattened_args.insert(to_insert.args.begin(), to_insert.args.end());
+        } else {
+            flattened_args.insert(a);
+        }
+    }
```

Comment:
You don't need two sets, `args` and `flattened_args` right? You can check whether `a` is a `caller` in Line 163 and insert the elements to `args`